src/app/shared/infinity-scroll/infinity-scroll.directive.ts

```diff
@@ -0,0 +1,42 @@
+import {Directive, EventEmitter, HostListener, Output} from '@angular/core';
+import {LoadDirection} from '../load-direction/load-direction.enum';
+
+@Directive({
+    selector: '[appInfinityScroll]',
+})
+export class InfinityScrollDirective {
+    @Output() readonly loadData = new EventEmitter<LoadDirection>();
+
+    private readonly borderOffset = 100;
+    private lastScrollTop = 0;
+
+    @HostListener('scroll', ['$event', '$event.target']) onScroll(
+        event: MouseEvent,
+        elem: HTMLElement,
+    ) {
+        this.borderObserver(elem);
+        event.stopPropagation();
+    }
+
+    private borderObserver(elem: HTMLElement) {
+        if (elem) {
+            const CURRENT_SCROLL_TOP = elem.scrollTop;
+            const HEIGHT = elem.getBoundingClientRect().height;
+            const SCROLL_HEIGHT = elem.scrollHeight;
+
+            if (
+                SCROLL_HEIGHT - (HEIGHT + CURRENT_SCROLL_TOP) <= this.borderOffset &&
+                this.lastScrollTop < CURRENT_SCROLL_TOP
+            ) {
+                this.loadData.emit(LoadDirection.fromBottom);
+            } else if (
+                CURRENT_SCROLL_TOP <= this.borderOffset &&
+                this.lastScrollTop > CURRENT_SCROLL_TOP
+            ) {
+                this.loadData.emit(LoadDirection.fromTop);
+            }
```

Comment:
На самом деле, я противник подхода с `else/if`, он крайне не удобен и не читаем. Лучшей практикой является ранний ретерн:
```suggestion
            const isScrollToBottom = this.lastScrollTop < CURRENT_SCROLL_TOP;
            const nearBottom = SCROLL_HEIGHT - (HEIGHT + CURRENT_SCROLL_TOP) <= this.borderOffset;
            
            if (nearBottom && isScrollToBottom) {
                this.loadData.emit(LoadDirection.fromBottom);
                
                return;
            }
            
            const nearTop = CURRENT_SCROLL_TOP <= this.borderOffset;
            
            if (nearTop && !isScrollToBottom) {
                this.loadData.emit(LoadDirection.fromTop);
                
                return;
            }
```

Reply:
Ничего не скажешь о подходе: понравился, не понравился?)

Reply:
норм, но мне больше понравилось подход с вынесением вычислений в "говорящие" константы

Reply:
Это уже немного другое) Тут поднимается вопрос про else/if VS ранний return

А выносить говорящие константы и там и там нужно) Но кажется, понял тебя